ixwebsocket/IXWebSocketTransport.cpp

```diff
@@ -176,33 +189,67 @@ namespace ix
         _onCloseCallback = onCloseCallback;
     }
 
-    // Only consider send time points for that computation.
-    // The receive time points is taken into account in Socket::poll (second parameter).
+    // Only consider send PING time points for that computation.
     bool WebSocketTransport::heartBeatPeriodExceeded()
     {
-        std::lock_guard<std::mutex> lock(_lastSendTimePointMutex);
+        std::lock_guard<std::mutex> lock(_lastSendPingTimePointMutex);
         auto now = std::chrono::steady_clock::now();
-        return now - _lastSendTimePoint > std::chrono::seconds(_heartBeatPeriod);
+        return now - _lastSendPingTimePoint > std::chrono::seconds(_heartBeatPeriod);
     }
 
-    void WebSocketTransport::poll()
+    bool WebSocketTransport::pongReceiveDelayExceeded()
     {
-        _socket->poll(
-            [this](PollResultType pollResult)
+        if (_heartBeatFactorDisconnectOnNoResponse == kDefaultHeartBeatFactorDisconnectOnNoResponse)
+            return false;
+
+        std::lock_guard<std::mutex> lock(_lastReceivePongTimePointMutex);
+        auto now = std::chrono::steady_clock::now();
+        return now - _lastReceivePongTimePoint > std::chrono::seconds(_heartBeatFactorDisconnectOnNoResponse*_heartBeatPeriod);
+    }
+
+    void WebSocketTransport::runHeartBeat()
+    {
+        using millis = std::chrono::duration<double, std::milli>;
+
+        while (_future.wait_for(std::chrono::milliseconds(0)) == std::future_status::timeout)
+        {
+            if(_readyState == OPEN)
             {
-                // If (1) heartbeat is enabled, and (2) no data was received or
-                // send for a duration exceeding our heart-beat period, send a
-                // ping to the server.
-                if (pollResult == PollResultType::Timeout &&
-                    heartBeatPeriodExceeded())
+                // if (1) disconnect on no PONG received is enabled and (2) heartbeat is enabled and (3) duration exceeded the maximum delay, then close the connection
+                if(pongReceiveDelayExceeded())
+                {
+                    close();
```

Comment:
I guess we need to close with some reason, maybe 1006 - Abnormal Closure.

Reply:
1006 seems to be for truncated payload and things like that, so I'm not sure it is for what we have done here

Reply:
ok, use some another code and not 1000.
Also set some reason to _closeReason